crates/solver/src/settlement/settlement_encoder.rs

```diff
@@ -654,4 +672,38 @@ pub mod tests {
 
         assert!(encoder0.merge(encoder1).is_err());
     }
+
+    #[test]
+    fn encoding_strips_unnecessary_tokens_and_prices() {
+        let prices = hashmap! {token(1) => 7.into(), token(2) => 2.into(),
+        token(3) => 9.into(), token(4) => 44.into()};
+
+        let mut encoder = SettlementEncoder::new(prices);
+
+        let order_1_2 = OrderBuilder::default()
+            .with_sell_token(token(1))
+            .with_sell_amount(33.into())
+            .with_buy_token(token(2))
+            .with_buy_amount(11.into())
+            .build();
+        encoder
+            .add_trade(order_1_2, 4.into(), 0.into(), false)
+            .unwrap();
+
+        let weth = dummy_contract!(WETH9, token(3));
+        encoder.add_unwrap(UnwrapWethInteraction {
+            weth,
+            amount: 12.into(),
+        });
+
+        let encoded = encoder.finish();
+
+        // only token 1 and 2 have been included in orders by traders
+        let expected_tokens: Vec<_> = [1, 2].into_iter().map(token).collect();
+        assert_eq!(expected_tokens, encoded.tokens);
+
+        // only the prices for token 1 and 2 remain and they are in the correct order
+        let expected_prices: Vec<_> = [7, 2].into_iter().map(U256::from).collect();
+        assert_eq!(expected_prices, encoded.clearing_prices);
+    }
```

Comment:
nit: Can we also assert that the trade's token indices are correct here?

Reply:
Done 👍